geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/SizeSeparatedBucketIndex.scala

```diff
@@ -0,0 +1,201 @@
+      throw new IllegalArgumentException(s"Envelope $envelope exceeds the max tier size ${sizes.last}")
+    }
+  }
+
+  private class Tier(val maxSizeX: Double,
+                     val maxSizeY: Double,
+                     buckets: Array[Array[ConcurrentHashMap[String, T]]],
+                     gridSnap: GridSnap) {
+
+    // we can safely use volatile instead of synchronized here, as this is a primitive boolean whose
+    // state doesn't depend on its own value
+    @volatile
+    var empty: Boolean = true
+
+    private val maxX = buckets.length - 1
+    private val maxY = buckets(0).length - 1
+
+    def bucket(x: Double, y: Double): ConcurrentHashMap[String, T] = buckets(snapX(x))(snapY(y))
+
+    // the bucket is selected based on the envelope centroid
+    def bucket(envelope: Envelope): ConcurrentHashMap[String, T] =
+      buckets(snapX((envelope.getMinX + envelope.getMaxX) / 2.0))(snapY((envelope.getMinY + envelope.getMaxY) / 2.0))
+
+    def iterator(xmin: Double, ymin: Double, xmax: Double, ymax: Double): Iterator[T] =
+      if (empty) { Iterator.empty } else { new TierIterator(xmin, ymin, xmax, ymax) }
+
+    def size(): Int = {
+      if (empty) { 0 } else {
+        var size = 0
+        var i = 0
+        while (i <= maxX) {
+          var j = 0
+          while (j <= maxY) {
+            size += buckets(i)(j).size()
+            j += 1
+          }
+          i += 1
+        }
+        size
+      }
+    }
+
+    def clear(): Unit = {
+      var i = 0
+      while (i <= maxX) {
+        var j = 0
+        while (j <= maxY) {
+          buckets(i)(j).clear()
+          j += 1
+        }
+        i += 1
+      }
+    }
+
+    private def snapX(x: Double): Int = {
+      val i = gridSnap.i(x)
+      if (i != -1) { i } else if (x < extents.getMinX) { 0 } else { maxX }
+    }
+
+    private def snapY(y: Double): Int = {
+      val j = gridSnap.j(y)
+      if (j != -1) { j } else if (y < extents.getMinY) { 0 } else { maxY }
+    }
+
+    /**
+      * Iterator over a range of buckets
+      */
+    class TierIterator (xmin: Double, ymin: Double, xmax: Double, ymax: Double) extends Iterator[T] {
+
+      private val maxi = snapX(xmax + maxSizeX)
+      private val minj = snapY(ymin - maxSizeY)
+      private val maxj = snapY(ymax + maxSizeY)
+
+      private var i = snapX(xmin - maxSizeX)
+      private var j = minj
+      private var iter = buckets(i)(j).values.iterator() // note: `.values` is a cached view
+
+      @tailrec
+      override final def hasNext: Boolean = iter.hasNext || {
+        if (i == maxi && j == maxj) { false } else {
+          if (j < maxj) {
+            j += 1
+          } else {
+            j = minj
+            i += 1
+          }
+          iter = buckets(i)(j).values.iterator() // note: `.values` is a cached view
+          hasNext
+        }
+      }
+
+      override def next(): T = iter.next()
+    }
+
+  }
+}
+
+object SizeSeparatedBucketIndex {
+  // TODO these are somewhat arbitrary
```

Comment:
Update to point to an actual ticket?

Also, I had good clean fun with the SSI only realizing some of the 'sizes', so I do like this approach.

Reply:
Configurations here can be a follow-on ticket.

Reply:
added a ticket